### ISSUE TITLE
Connection closed too early

### ISSUE THREAD

**Issue by zvin** [resin-io/sshproxy#30]
https://app.frontapp.com/open/cnv_6s4t5l
https://www.flowdock.com/app/rulemotion/resin-tech/threads/fMzdZ0HIyA8TZYXTTZNGU9fs7w2
https://www.flowdock.com/app/rulemotion/resin-tech/threads/FCOgP-6bYRwFn_tRzs3wWgHHD7M

<img src="https://frontapp.com/assets/img/icons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_736p)

**Comment by zvin** [comment]
Trying to clone from resin git sometimes fails with
```
fatal: The remote end hung up unexpectedly
fatal: early EOF
fatal: index-pack failed
```

resin-git logs contain
```
23 Jan 2018 20:35:45.409920888a69906 sshproxy[26011]: Closed SSH channel with 95.16.11.188:13376
23 Jan 2018 20:35:45.673920888a69906 sshproxy[26011]: ERROR: close tcp 172.17.0.1:22->95.16.11.188:13376: use of closed network connection
```

**Comment by zvin** [comment]
We close the connection before we finish handling it.
Sometimes this defer https://github.com/resin-io/sshproxy/blob/master/sshproxy.go#L145 is called too early because hanldeChannels returns before finishing hanling the last channel.

In handleChannels we have a go s.handleRequests which means that we will return before finishing handleRequests and the defer from upgradeConnection will be called and the connection will be closed.
